tools/restatectl/src/connection.rs

```diff
@@ -0,0 +1,343 @@
+        // a total of 50% + 1 nodes.
+
+        self.get_latest(
+            try_nodes.into_iter(),
+            (cluster_size / 2) + 1,
+            MetadataKind::Logs,
+            guard,
+            |ident| Version::from(ident.logs_version),
+        )
+        .await
+    }
+
+    /// Gets NodesConfiguration object. This function will try `try_best_of` nodes from the provided
+    /// addresses. Worst case all addresses will be tried if some or all nodes did not satisfy the query
+    async fn get_latest<T, M>(
+        &self,
+        addresses: impl Iterator<Item = &AdvertisedAddress>,
+        mut try_best_of: usize,
+        kind: MetadataKind,
+        mut guard: MutexGuard<'_, Option<T>>,
+        version_map: M,
+    ) -> Result<T, ConnectionInfoError>
+    where
+        T: StorageDecode + Versioned + Clone,
+        M: Fn(&IdentResponse) -> Version,
+    {
+        if let Some(meta) = &*guard {
+            return Ok(meta.clone());
+        }
+
+        let mut latest_meta: Option<T> = None;
+        let mut answer = false;
+        let mut errors = NodesErrors::default();
+        let mut cache = self.cache.lock().await;
+
+        let request = GetMetadataRequest {
+            kind: kind.into(),
+            sync: false,
+        };
+
+        for address in addresses {
+            let channel = cache.entry(address.clone()).or_insert_with(|| {
+                debug!("connecting to {address}");
+                grpc_channel(address.clone())
+            });
+
+            let mut client = NodeCtlSvcClient::new(channel.clone())
+                .accept_compressed(CompressionEncoding::Gzip)
+                .send_compressed(CompressionEncoding::Gzip);
+
+            let response = match client.get_ident(()).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address.clone(), status);
+                    continue;
+                }
+            };
+
+            // node is reachable and has answered
+
+            answer = true;
+            if response.status != NodeStatus::Alive as i32 {
+                // node did not join the cluster yet.
+                continue;
+            }
+
+            if version_map(&response)
+                <= latest_meta
+                    .as_ref()
+                    .map(|c| c.version())
+                    .unwrap_or(Version::INVALID)
+            {
+                // has older version than we have.
+                continue;
+            }
+
+            let mut response = match client.get_metadata(request).await {
+                Ok(response) => response.into_inner(),
+                Err(status) => {
+                    errors.error(address.clone(), status);
+                    continue;
+                }
+            };
+
+            let meta = StorageCodec::decode::<T, _>(&mut response.encoded).map_err(|err| {
+                ConnectionInfoError::InvalidNodesConfiguration(address.clone(), err)
```

Comment:
Is this always the `NodesConfiguration` we are deserializing here?

Reply:
Ah, definitely not. Thank you for catching this one